templates/docs/examples/patterns/modal/_script.js

```diff
@@ -1,22 +1,119 @@
+      lastFocus = document.activeElement;
+    }
+  }
+
+  // Attempts to focus given element
+  function attemptFocus(child) {
+    if (child.focus) {
+      ignoreFocusChanges = true;
+      child.focus();
+      ignoreFocusChanges = false;
+      return document.activeElement === child;
     }
+
+    return false;
+  }
+
+  // Focuses first child element
+  function focusFirstDescendant(element) {
+    for (var i = 0; i < element.childNodes.length; i++) {
+      var child = element.childNodes[i];
+      if (attemptFocus(child) || focusFirstDescendant(child)) {
+        return true;
+      }
+    }
+    return false;
   }
-}
 
-// Add click handler for clicks on elements with aria-controls
-document.addEventListener('click', function (event) {
-  var targetControls = event.target.getAttribute('aria-controls');
-  if (targetControls) {
-    toggleModal(document.getElementById(targetControls));
+  // Focuses last child element
+  function focusLastDescendant(element) {
+    for (var i = element.childNodes.length - 1; i >= 0; i--) {
+      var child = element.childNodes[i];
+      if (attemptFocus(child) || focusLastDescendant(child)) {
+        return true;
+      }
+    }
+    return false;
   }
-});
+
+  /**
+    Toggles visibility of modal dialog.
+    @param {HTMLElement} modal Modal dialog to show or hide.
+    @param {HTMLElement} sourceEl Element that triggered toggling modal
+    @param {Boolean} open If defined as `true` modal will be opened, if `false` modal will be closed, undefined toggles current visibility.
+  */
+  function toggleModal(modal, sourceEl, open) {
+    if (modal && modal.classList.contains('p-modal')) {
+      if (typeof open === 'undefined') {
+        open = modal.style.display === 'none';
+      }
+
+      if (open) {
+        currentDialog = modal;
+        modal.style.display = 'flex';
+        focusFirstDescendant(modal);
+        focusAfterClose = sourceEl;
+        document.addEventListener('focus', trapFocus, true);
+      } else {
+        modal.style.display = 'none';
+        if (focusAfterClose && focusAfterClose.focus) {
+          focusAfterClose.focus();
+        }
+        document.removeEventListener('focus', trapFocus, true);
+        currentDialog = null;
+      }
+    }
+  }
+
+  // Find and hide all modals on the page
+  function closeModals() {
+    var modals = [].slice.apply(document.querySelectorAll('.p-modal'));
+    modals.forEach(function (modal) {
+      toggleModal(modal, false);
+    });
+  }
+
+  // Add click handler for clicks on elements with aria-controls
+  document.addEventListener('click', function (event) {
+    var targetControls = event.target.getAttribute('aria-controls');
+    if (targetControls) {
+      toggleModal(document.getElementById(targetControls), event.target);
+    }
+  });
+
+  // Add handler for closing modals using ESC key.
+  document.addEventListener('keydown', function (e) {
+    e = e || window.event;
+
+    if (e.code === 'Escape') {
+      closeModals();
```

Comment:
@bartaz when I am on the page, no matter where the focus is, I can open and close the modal by tapping the escape key - because this `closeModals` is toggling, rather than closing, any modal on the page. I think the `toggleModal` call on [line 94](https://github.com/canonical-web-and-design/vanilla-framework/pull/3712/files#diff-3ef8df5bbb32785f7b762e8c3413d297836863be06e9b001c6030314779bab27R94) needs to pass in a third argument i.e. `toggleModal(modal, false, false);`, otherwise the if statement on [line 69](https://github.com/canonical-web-and-design/vanilla-framework/pull/3712/files#diff-3ef8df5bbb32785f7b762e8c3413d297836863be06e9b001c6030314779bab27R69) will always set the `open` variable to be the opposite of whatever the modal currently is.

Reply:
ah, nice catch, thanks!

Reply:
Right, that got broken when I added third argument to `toggleModal`.